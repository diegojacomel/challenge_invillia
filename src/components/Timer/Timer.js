/** Modules */
import React, { Component } from 'react';
import styled from 'styled-components';

const TimerStyle = styled.div`
    font-weight: bold;
    font-size: ${props => props.theme.fontSize.hg};
`

const TimerElement = ({ children }) => (
    <TimerStyle>
        {children}
    </TimerStyle>
)

class Timer extends Component {
    constructor() {
        super();

        this.state = {
            time: {},
            seconds: 120
        };
        this.timer = 0;
        this.startTimer = this.startTimer.bind(this);
        this.countDown = this.countDown.bind(this);
    }

    componentDidMount() {
        let timeLeftVar = this.secondsToTime(this.state.seconds);
        this.setState({ time: timeLeftVar });

        this.startTimer()
    }

    componentDidUpdate() {
        const { state, props: { endTime } } = this;

        if (+state.seconds === 0) {
            endTime()
        }
    }

    secondsToTime(secs) {
        let divisor_for_minutes = secs % (60 * 60);
        let minutes = Math.floor(divisor_for_minutes / 60);

        let divisor_for_seconds = divisor_for_minutes % 60;
        let seconds = Math.ceil(divisor_for_seconds);

        if (minutes < 10) {
            minutes = `0${minutes}`
        }

        if (seconds < 10) {
            seconds = `0${seconds}`
        }

        let obj = {
            "m": minutes,
            "s": seconds
        };

        return obj;
    }

    startTimer() {
        if (this.timer === 0 && this.state.seconds > 0) {
            this.timer = setInterval(this.countDown, 1000);
        }
    }

    countDown() {
        // Remove one second, set state so a re-render happens.
        let seconds = this.state.seconds - 1;
        this.setState({
            time: this.secondsToTime(seconds),
            seconds: seconds,
        });

        // Check if we're at zero.
        if (seconds === 0) {

            clearInterval(this.timer);
        }
    }

    render() {
        const { state, props: { endTime } } = this;

        return (
            <TimerElement id="testTimer" endTime={endTime}>
                {state.time.m}:{state.time.s}
            </TimerElement>
        );
    }
}

export default Timer;