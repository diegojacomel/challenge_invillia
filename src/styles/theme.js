const primaryTheme = {
    colors: {
        primary: '#007bff',
        success: '#28a745',
        secondary: '#6c757d',
        warning: "#ffc107",
        black: '#000000',
        white: '#FFFFFF',
        gray: '#cccccc',
        grayLight: '#F2F2F2',
        grayDark: '#333333'
    },
    fontSize: {
        xs: "14px",
        sm: "16px",
        md: "20px",
        lg: "25px",
        hg: "40px",
        hhg: "60px"
    },
    spacing: {
        none: '0',
        xxs: '3px',
        xs: '5px',
        ss: '8px',
        sm: '10px',
        ssm: '15px',
        md: '20px',
        mmd: '25px',
        lg: '30px',
        llg: '35px',
        xg: '40px',
        xxg: '50px',
        gt: '80px',
        ggt: '100px',
        hg: '200px',
        hhg: '400px',
        quarter: '25%',
        middle: '50%',
        full: '100%'
    },
    doubleSpacing: {
        xs: "2px 5px",
        sm: "3px 10px",
        md: "5px 12px",
        lg: "8px 16px",
        xg: "10px 20px",
        hg: "20px 40px"
    },
    rounded: {
        none: '0',
        xs: '2px',
        sm: '3px',
        md: '5px',
        lg: '10px'
    }
}

export {
    primaryTheme
}