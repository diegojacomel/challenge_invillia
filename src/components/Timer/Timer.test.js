// Modules
import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

// Theme
import { primaryTheme } from '../../styles/theme';

// Component
import Timer from './Timer';

describe('Timer', () => {
    const wrapper = shallow(<Timer />);

    test('Render without crashing', () => {
        shallow(<Timer />);
    })

    test('It exist', () => {
        expect(wrapper.find('#testTimer')).toExist();
    })

    test('At least one', () => {
        expect(wrapper).toContainMatchingElement('#testTimer');
    })

    test('Have display name', () => {
        expect(wrapper.find('#testTimer')).toHaveDisplayName('TimerElement');
    })

    test('Match selector', () => {
        expect(wrapper.find('#testTimer')).toMatchSelector('TimerElement');
    })

    it('Render correctly', () => {
        const tree = renderer.create(
            <ThemeProvider theme={primaryTheme}>
                <Timer />
            </ThemeProvider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    })
})