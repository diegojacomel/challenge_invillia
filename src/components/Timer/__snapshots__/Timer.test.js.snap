// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer Render correctly 1`] = `
<div
  className="sc-bdVaJa gOLQNu"
>
  02
  :
  00
</div>
`;
